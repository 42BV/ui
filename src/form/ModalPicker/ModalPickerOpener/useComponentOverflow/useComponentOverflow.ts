import { RefObject, useCallback, useLayoutEffect, useState } from 'react';
import { useResizeObserver } from '../useResizeObserver/useResizeObserver';

export function useComponentOverflow(
  ref: RefObject<HTMLElement>,
  values: React.ReactNode
) {
  const [componentOverflow, setComponentOverflow] = useState(
    isOverflowing(ref.current)
  );

  const handleResize = useCallback(
    handleResizeCallback(ref, setComponentOverflow),
    [ref]
  );

  const resizeObserver = useResizeObserver(handleResize);

  useLayoutEffect(layoutEffect(ref, handleResize, resizeObserver), [
    ref,
    handleResize,
    values,
    resizeObserver
  ]);

  return componentOverflow;
}

export function isOverflowing(el: HTMLElement | null) {
  return el ? el.clientWidth < el.scrollWidth : false;
}

export function handleResizeCallback(
  ref: RefObject<HTMLElement>,
  setComponentOverflow: (value: boolean) => void
) {
  return () => {
    if (ref.current) {
      setComponentOverflow(isOverflowing(ref.current));
    }
  };
}

export function layoutEffect(
  ref: RefObject<HTMLElement>,
  handleResize: () => void,
  resizeObserver?: any
) {
  return () => {
    if (!ref.current) {
      return;
    }

    handleResize();

    const el = ref.current;

    if (resizeObserver) {
      resizeObserver.observe(el);

      return () => {
        resizeObserver.unobserve(el);
      };
    }

    el.addEventListener('resize', handleResize, { passive: true });

    return () => {
      el.removeEventListener('resize', handleResize);
    };
  };
}
