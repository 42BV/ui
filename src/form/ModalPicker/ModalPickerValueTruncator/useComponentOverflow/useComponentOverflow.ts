import { RefObject, useCallback, useLayoutEffect, useState } from 'react';
import { useResizeObserver } from '../useResizeObserver/useResizeObserver';

export function useComponentOverflow<T>(
  ref: RefObject<HTMLElement>,
  values: T | T[]
): boolean {
  const [ componentOverflow, setComponentOverflow ] = useState(
    isOverflowing(ref.current)
  );

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const handleResize = useCallback(
    handleResizeCallback(ref, setComponentOverflow),
    [ ref ]
  );

  const resizeObserver = useResizeObserver(handleResize);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useLayoutEffect(layoutEffect(ref, handleResize, resizeObserver), [
    ref,
    handleResize,
    values,
    resizeObserver
  ]);

  return componentOverflow;
}

export function isOverflowing(el: HTMLElement | null): boolean {
  return el ? el.clientWidth < el.scrollWidth : false;
}

export function handleResizeCallback(
  ref: RefObject<HTMLElement>,
  setComponentOverflow: (value: boolean) => void
): () => void {
  return () => {
    if (ref.current) {
      setComponentOverflow(isOverflowing(ref.current));
    }
  };
}

export function layoutEffect(
  ref: RefObject<HTMLElement>,
  handleResize: () => void,
  resizeObserver?: any
): () => (() => void) | void {
  return () => {
    if (!ref.current) {
      return;
    }

    handleResize();

    const el = ref.current;

    if (resizeObserver) {
      resizeObserver.observe(el);

      return () => {
        resizeObserver.unobserve(el);
      };
    }

    el.addEventListener('resize', handleResize, { passive: true });

    return () => {
      el.removeEventListener('resize', handleResize);
    };
  };
}
