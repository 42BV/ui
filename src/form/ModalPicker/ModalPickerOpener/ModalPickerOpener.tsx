import React, { useRef } from 'react';
import { Button } from 'reactstrap';
import Tooltip from '../../../core/Tooltip/Tooltip';
import { useComponentOverflow } from './useComponentOverflow/useComponentOverflow';

interface Props {
  /**
   * Function to open the modal, called when the button is clicked.
   */
  openModal: () => void;

  /**
   * The label to display on the button.
   */
  label: React.ReactNode;

  /**
   * The display of selected item(s).
   */
  values?: React.ReactNode;
}

export function ModalPickerOpener(props: Props) {
  const { openModal, label, values } = props;

  const ref = useRef<HTMLElement>(null);

  const isOverflowing = useComponentOverflow(ref, values);

  return (
    <div className="d-flex align-items-center">
      {values ? (
        <span className="text-truncate mr-1 position-relative" ref={ref}>
          {isOverflowing ? (
            <Tooltip
              content={values}
              className="w-100 h-100 position-absolute"
              tag="div"
            >
              {' '}
            </Tooltip>
          ) : null}
          {values}
        </span>
      ) : null}
      <Button
        color="primary"
        onClick={openModal}
        className="flex-grow-0 flex-shrink-0"
      >
        {label}
      </Button>
    </div>
  );
}
