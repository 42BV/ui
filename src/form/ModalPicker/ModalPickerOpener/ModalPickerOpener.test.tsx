import React from 'react';
import { shallow } from 'enzyme';
import { ModalPickerOpener } from './ModalPickerOpener';
import { adminUser } from '../../../test/fixtures';
import toJson from 'enzyme-to-json';
import * as useComponentOverflow from './useComponentOverflow/useComponentOverflow';

describe('Component: ModalPickerOpener', () => {
  function setup({
    hasValues = false,
    withTooltip = false
  }: {
    hasValues?: boolean;
    withTooltip?: boolean;
  }) {
    const openModalSpy = jest.fn();
    const values = hasValues ? adminUser.email : undefined;

    jest
      .spyOn(useComponentOverflow, 'useComponentOverflow')
      .mockImplementation(() => withTooltip);

    const modalPickerOpener = shallow(
      <ModalPickerOpener
        openModal={openModalSpy}
        label="Best friend"
        values={values}
      />
    );

    return { modalPickerOpener, openModalSpy };
  }

  describe('ui', () => {
    test('without values', () => {
      const { modalPickerOpener } = setup({});
      expect(toJson(modalPickerOpener)).toMatchSnapshot(
        'Component: ModalPickerOpener => ui => without values'
      );
    });

    test('with values', () => {
      const { modalPickerOpener } = setup({ hasValues: true });
      expect(toJson(modalPickerOpener)).toMatchSnapshot(
        'Component: ModalPickerOpener => ui => with values'
      );
    });

    test('tooltip', () => {
      const { modalPickerOpener } = setup({
        hasValues: true,
        withTooltip: true
      });

      expect(toJson(modalPickerOpener)).toMatchSnapshot(
        'Component: ModalPickerOpener => ui => with tooltip'
      );
    });
  });

  describe('events', () => {
    it('should open the modal when the select button is clicked', () => {
      const { modalPickerOpener, openModalSpy } = setup({});

      // @ts-ignore
      modalPickerOpener
        .find('Button')
        .props()
        // @ts-ignore
        .onClick();

      expect(openModalSpy).toHaveBeenCalledTimes(1);
    });
  });
});
