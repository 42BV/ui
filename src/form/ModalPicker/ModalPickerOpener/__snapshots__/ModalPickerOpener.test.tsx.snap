// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ModalPickerOpener ui tooltip: Component: ModalPickerOpener => ui => with tooltip 1`] = `
<div
  className="d-flex align-items-center"
>
  <span
    className="text-truncate mr-1 position-relative"
  >
    <Tooltip
      className="w-100 h-100 position-absolute"
      content="admin@42.nl"
      tag="div"
    >
       
    </Tooltip>
    admin@42.nl
  </span>
  <Button
    className="flex-grow-0 flex-shrink-0"
    color="primary"
    onClick={[MockFunction]}
    tag="button"
  >
    Best friend
  </Button>
</div>
`;

exports[`Component: ModalPickerOpener ui with values: Component: ModalPickerOpener => ui => with values 1`] = `
<div
  className="d-flex align-items-center"
>
  <span
    className="text-truncate mr-1 position-relative"
  >
    admin@42.nl
  </span>
  <Button
    className="flex-grow-0 flex-shrink-0"
    color="primary"
    onClick={[MockFunction]}
    tag="button"
  >
    Best friend
  </Button>
</div>
`;

exports[`Component: ModalPickerOpener ui without values: Component: ModalPickerOpener => ui => without values 1`] = `
<div
  className="d-flex align-items-center"
>
  <Button
    className="flex-grow-0 flex-shrink-0"
    color="primary"
    onClick={[MockFunction]}
    tag="button"
  >
    Best friend
  </Button>
</div>
`;
